Old->New index migration

poltak opened the pull request:
Addresses #363

Current status:

Migration will start when the browser is idle. Whenever the user comes back, it will save its progress (in-mem and persisted) and stop. Once idle again, migration will resume from saved progress point.

Once finished, it will:
- show an interactive notification (currently says "Your data has been migrated to our new and improved index!" and requires interaction)
- swap the search backend to use the new index (old index will stop being used for all main functionality)
- set saved progress state to special value (migration is a one-time thing - once finished, we don't want it starting again)

There is no current visual indication of this happening, apart from the notification upon completion.

Working quite well in a number of contrived tests I've done with 100s of MBs of real data.

Some TODOs:

- [x] ensure it doesn't try to start, or start and instantly complete, without any old index data (new/post-update users)
- [ ] figure out what to do with the old index data after complete
  -  I think in #363 I talked about immediately deleting it once migration is complete, but it feels scary in case there's a bug with the iteration and the `onComplete` logic gets prematurely run - maybe we wait a week or so after the rollout to see if we need to address anything first?
- [x] write some instructions on how to manually test; it's a bit tricky
  - how to force a dev build to use the old index to collect some real data
  - how to upgrade and know if migration is running
  - how to tell if search is using old or new index as backend
- [ ] ideas for less sucky automated tests? really important we get this feature working correctly
  - current tests just test migrating a single page and checking data integrity, but a single page is not really realistic usage of this feature (most users data expected to be in 100s of MBs)

poltak commented:
Here's the (slightly painful) instructions for trying out migration:

1. set up ext to use old implementation as index backend __(+ build)__
    - go to [`src/search/migration/index.ts` module](https://github.com/WorldBrain/Memex/blob/feature/index-migration/src/search/migration/index.ts#L18-L25). Comment out lines 18-25 (subject to change; basically it's the stuff inside the `searchIndex.dataReady.then(() => { ... }`)
    - add this as a new line under the commented-out stuff:
 `searchIndex.dataReady.then(() => (searchIndex.useOld = true))`
2. go get some data, via imports and/or page visits
3. discard all those changes you made in step 1 __(+ build)__
4. do an extension reload to update your running extension to the new build
5. migration should start after ~1 min of not touching your computer

 As in the OP, there's no visual indication that things are running until the notification once migration completes. If you want to test the current implementation, I would advise opening the Chrome dev tools for the background script, go to Application tab, then in the sidebar click `Application -> Clear storage`. This will open up the page with your storage quota usage which, if you leave that open and don't touch anything, should remain mostly static until migration starts, where it will start going up and down in little bits until finished (or interrupted it by resuming activity - which is another thing to try out).

Once finished, the next time you use any ext functionality - search, page visit, etc. - things should automatically be moved over and now using the new implementation. Old IDB instances should still exist (called `IDBWrapper-worldbrain-terms` and `_pouch_webmemex` in the Applications dev tools tab), but just never used. 
Again, there's no visual indicator to tell which index backend is in-use, but you could write `index.useOld` in the background script dev tools console. It should print `true` if everything is still using the old implementation, or `false` for new. (compare pre and post migration)

oliversauter commented:
Ok, so ran a couple of tests, worked smoothly:
1) Migration itself went smoothly
   - My last try with a larger set of about 3000 documents (96mb) took about 4 minutes. That seemed odd/long, will try it out tomorrow again. 

4) If the migration is interruptable/resumable, do we need to have a 1 minute as the threshold? Can we make it shorter, or are there some dangers in doing that? 
5) how do we check on the integrity of the data? To ensure everything went smoothly, do we just wait until problems come up and if that is the case, we still have the old data and could redo the migration?

I downloaded about 3k documents in the old index version, would be good if I could reuse this set when testing a couple of times (without importing them again). Is there a way to reset some stat so it just reruns the migration on extension reload?

mukeshkharita commented:
@oliversauter @poltak I tested on the 1100 pages(70mb), It worked smoothly and it took less than 2 minutes.

ShishKabab commented:
Do we have any idea of medium data set sizes and storage quotas for users? Would suck if data consumption gets to high due to the migration and the browser decides to delete all of our users' data  ;)  Not very probable, just being paranoid  :)

ShishKabab commented on src/search/migration/import-export.test.ts:
```diff
@@ -0,0 +1,159 @@
+        }
+
+        test('Importing data to new index', async () => {
+            index.useOld = false
```
When I'm importing my data, isn't useOld still set to true?

poltak replied:
Yes, actually the `import` code just uses Dexie directly now (to avoid a lot of unneeded processing our index interface methods, like `addNewPage`, do), so the `useOld` value is not used. This line is meant for the following `index.search` call, so the search happens on the freshly migrated new index. Very confusing here, so will move it.

ShishKabab commented on src/search/migration/migration-manager.ts:
```diff
@@ -0,0 +1,138 @@
+        this.isCancelled = false
+        await MigrationManager.persistProgressState(this.currKey)
+
+        throw new Error()
```
Could this be an Error subclass with a special property (.type, .paused, etc.) to more easily handle it differently than other errors?

poltak replied:
Great suggestion!

ShishKabab commented:
Awesome stuff here! You did a great job of splitting up this complexity into understandable pieces of code. Ignored a few minor cosmetic things because this code is temporary anyway. Do you still see it possible to re-use lots of this code for import/export functionality to files / Google Drive?

oliversauter commented:
Ok another round of testing: 
3200 docs with 70MB took about 5 minutes and worked smoothly. 

I would though reduce the time to activate idle mode to 20 seconds. This way the process can start multiple times and is over pretty fast. a 1 minute waiting time, including a roughly 20-25minute migration time seems really long and may not pull everyone over in a reasonable time. 

Does the migration process kickoff interfere with the stats we send? Can it come to complications between the two?

poltak commented:
@oliversauter another thing to think about: at the end of migration it currently just shows the notification. On Chrome it will stay there so the user sees it, but FF users could very well not know the migration ever happened (FF notifs are timeout-only). Maybe that's a not a bad thing though; ideally migration should be seamless from a user POV and shouldn't really require them to know about it. What do you think? We could also open a new tab to an info post about the migration, although more intrusive.

@ShishKabab forgot to reply to this:

> Do we have any idea of medium data set sizes and storage quotas for users? Would suck if data consumption gets to high due to the migration and the browser decides to delete all of our users' data 

We don't have any stats about user data sizes or storage quotas. However, if data exceeds quotas from the migration (or from any indexing in Memex) all the write ops will start throwing `QuotaExceededError`s, and only things like search should work until the user increases their quota. 
**Note we've never tested this** - not sure how to apart from filling up disk with a lot of crap, as the storage manager APIs don't seem available from web ext. 

The browser should not delete any of the data as we're using persistent storage (good info in this article: https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria).

oliversauter commented:
> Maybe that's a not a bad thing though

I think it would be good to make that notification, so we can show that we achieve things. 
I will write a new article in the knowledge base which you can link to. 
https://worldbrain.helprace.com/i57-super-fast-searching-is-now-live-to-all-users-import-speed-15x-faster

The text could be: 
"Your Memex has been upgraded to be 15x faster than before. Learn more."

Apart from that i think it is merge ready. Tested everything once again and runs smoothly. I think the 20s upgrade was a good one.

poltak commented:
@oliversauter I updated the notif to link to that article and updated the text. However it would still be highly unlikely that FF users will see it: migration will complete while idle, and the notif will only be shown for a few seconds.

What I've done is move the notification to be shown when the user comes back if migration finished while they were idle. So you might go away from the computer for 10mins and come back, move the mouse then the notif will show. Seems to be working okay and means more users should be aware.
